A98-debug-protocol.md

```diff
@@ -0,0 +1,224 @@
+It suffers from being relatively rigid in the topology it presents, and the set of node types available.
+
+
+### Related Proposals: 
+* A14 - channelz.
+
+## Proposal
+
+A new debug service proto will be added, `debug/debug.proto` in the `grpc.debug.v1` namespace.
+
+### Entities
+
+The fundamental building block of the protocol is a queryable entity.
+An entity describes one live object inside of a gRPC implementation.
+This might be a call, a channel, a transport, or a TCP socket. It might also describe a resource quota or a watched configuration file.
+
+One entity is described as such:
+
+```
+message Entity {
+    // The identifier for this entity.
+    int64 id = 1;
+    // The kind of this entity.
+    string kind = 2;
+    // Parents for this entity.
+    repeated int64 parents = 3;
+    // Instantaneous data for this entity.
+    repeated google.protobuf.Any data = 4;
+    // Historical trace information for the entity.
+    repeated TraceEvent trace = 5;
+}
+```
+
+Entities have state and configuration, and tend be be relatively long lived - such that querying them makes sense.
+
+**id**: An entity is identified by an id.
+These ids are allocated sequentially per the rationale in A14, and implementations should use the same id space for debug entities and channelz objects.
+
+**kind**: An entity has a kind. This is a string descriptor of the general category of object that this entity describes.
+We use strings here rather than enums so that implementations are free to extend the kind space with their own objects.
+Common entity kinds will see some level of standardization across stacks, but we expect many kinds to be specific per gRPC implementation.
+Initially implementations are expected to match kinds with channelz object types:
+kind `"channel"` -> `channelz.v1.Channel`, `"subchannel"` -> `channelz.v1.Subchannel`, `"server"` -> `channelz.v1.Server`, `"socket"` -> `channelz.v1.Socket`.
+
+
+**parents**: It's useful to be able to associate entities in parent/child hierarchies.
+For example, a channel has many subchannel children.
+Channelz listed specific kinds of children in its various node types - this tracking (and the need to produce it when sending an object) has caused contention issues in implementations in the past.
+Should the list of children (optionally of a particular kind) for an entity be desired - and that's common - a separate paginated service call can be made.
+Instead, this protocol lists only parents (as that set is far more stable).
+Multiple parents are allowed - to handle at least the case of C++ subchannels being owned by multiple channels.
+
+**data**: This is a list of protobuf Any objects describing the current state of this entity.
+Implementations may define their own protobufs to be exposed here, and common sets will be standardized separately.
+
+**trace**: Finally, an entity may supply a small summary of its history as a trace.
```

Comment:
Should we note here that traces can be either proactively recorded, in which case they are returned here, or recorded on-demand via the QueryTrace RPC method?

Should we impose any cap on the number of proactively recorded traces?

Reply:
Added some commentary